client/chat/channel.tsx

```diff
@@ -182,140 +178,121 @@ interface UserListProps {
+        throw new Error('Asked to render nonexistent user: ' + index)
       }
 
-      i -= offline.size + 1
-    }
-
-    if (i === 0) {
-      return <UserListPadding style={style} key={index} />
-    }
-
-    throw new Error('Asked to render nonexistent user: ' + index)
-  }
-
-  render() {
-    const { active, idle, offline } = this.props.users
-    const rowCount =
-      1 + active.size + (idle.size ? 1 : 0) + idle.size + (offline.size ? 1 : 0) + offline.size + 1
+      const idleHeaderIndex = idle.size ? active.size + 1 : null
+      let offlineHeaderIndex = null
+      if (offline.size) {
+        offlineHeaderIndex =
+          idleHeaderIndex != null ? idleHeaderIndex + idle.size + 1 : active.size + 1
+      }
 
-    return (
-      <UserListContainer ref={this._contentRef}>
-        <WindowListener event='resize' listener={this.updateDimensions} />
-        <UsersVirtualizedList
-          ref={this._listRef}
-          width={this.state.width}
-          height={this.state.height}
-          rowCount={rowCount}
-          rowHeight={this.getRowHeight}
-          rowRenderer={this.renderRow}
-        />
-      </UserListContainer>
-    )
-  }
+      switch (index) {
+        case 0:
+          return (
+            <UserListOverline style={style} key={index}>
+              Active ({active.size})
+            </UserListOverline>
+          )
+        case idleHeaderIndex:
+          return (
+            <UserListOverline style={style} key={index}>
+              Idle ({idle.size})
+            </UserListOverline>
+          )
+        case offlineHeaderIndex:
+          return (
+            <UserListOverline style={style} key={index}>
+              Offline ({offline.size})
+            </UserListOverline>
+          )
+        default:
+          let username: string | undefined
+          let faded = false
+          if (index < active.size + 1) {
+            username = active.get(index - 1)!
+          } else if (offlineHeaderIndex && index > offlineHeaderIndex) {
+            faded = true
+            username = offline.get(index - offlineHeaderIndex - 1)!
+          } else {
+            username = idleHeaderIndex ? idle.get(index - idleHeaderIndex - 1)! : undefined
+          }
+
+          if (username) {
+            return <UserListEntry style={style} username={username} key={username} faded={faded} />
+          }
+          throw new Error('Asked to render nonexistent user: ' + index)
+      }
+    },
+    [active, idle, offline, rowCount],
+  )
 
-  updateDimensions = () => {
-    const width = this._contentRef.current?.clientWidth ?? 0
-    const height = this._contentRef.current?.clientHeight ?? 0
-    if (this.state.width !== width || this.state.height !== height) {
-      this.setState({ width, height })
-    }
-  }
-}
+  return (
+    <UserListContainer ref={containerCallback}>
+      <UsersVirtualizedList
+        ref={listRef}
+        style={{ height: 'auto', display: 'inline-block', overflow: 'unset' }}
```

Comment:
I feel like this probably isn't necessary. I did this on the ladder table because we want the scrollbar to be on the root of the content, rather than on the table, but the Users list has its own scrollbar anyway.

Reply:
:+1: I added in a `overflowX: 'hidden'` to hide the scrollbar that appears.